chainer/testing/matrix.py

```diff
@@ -0,0 +1,56 @@
+import numpy
+
+from chainer.utils import argument
+
+
+def generate_matrix(shape, dtype=float, **kwargs):
+    r"""generate_matrix(shape, dtype=float, *, singular_values)
+
+    Generates a random matrix with given singular values.
+
+    This function generates a random NumPy matrix (or a stack of matrices) that
+    has specified singular values. It can be used to generate the inputs for a
+    test that can be instable when the input value behaves bad.
+
+    Notation: denote the shape of the generated array by :math:`(B..., M, N)`,
+    and :math:`K = min\{M, N\}`. :math:`B...` may be an empty sequence.
+
+    Args:
+        shape (tuple of int): Shape of the generated array, i.e.,
+            :math:`(B..., M, N)`.
+        dtype: Dtype of the generated array.
+        singular_values (array-like): Singular values of the generated
+            matrices. It must be broadcastable to shape :math:`(B..., K)`.
```

Comment:
Another possible spec is "of shape `(B'..., K)` s.t. `B'...` is broadcastable to `B...`".

Reply:
What would be the reason to prohibit broadcasting along the last axis?

Reply:
IMO, it is nice to allow it to generate orthogonal matrices.

A reason to prohibit it might be that NumPy's broadcast rule is not applied to the contracted axes of inputs of `numpy.matmul` and `numpy.linalg.solve`, (but it is applied to non-`'...'` axes of inputs of `numpy.einsum`).

Reply:
I see. I'm not sure if the prohibition in `matmul` and `solve` is essential (`matmul` case looks like a spec of gufunc), but that would make sense because it may be confusing to broadcast the contraction axes without specifying the contraction shape. In the case of `generate_matrix`, I hope it is not confusing because the full shape is explicitly given as an argument...